ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Wallets.hs

```diff
@@ -175,3 +200,25 @@ updateHdWalletAssurance
     -> IO (Either HD.UnknownHdRoot HdRoot)
 updateHdWalletAssurance pw hdRootId assuranceLevel =
     update' (pw ^. wallets) (UpdateHdWalletAssurance hdRootId assuranceLevel)
+
+{-------------------------------------------------------------------------------
+  Automatic creation of Wallet Account and Address
+-------------------------------------------------------------------------------}
+
+addAddress 
+    :: PassiveWallet
+    -> WalletId
+    -> PassPhrase
+    -> HdRoot
+    -> IO (Either CreateWalletError HdRoot)
+addAddress pw walletId passphrase hdRoot = do 
```

Comment:
Not sure why you need to take `HdRoot` as a parameter, only to return it in case of success. It would probably be better to return `IO (Either CreateWalletError ())` instead.

Also, you may want to use `ExceptT`. See examples in `Accounts` and `Addresses`.